framework/src/eventloop/eventloop_timer.c

```diff
@@ -24,210 +24,251 @@
+	sq_rem((FAR sq_entry_t *)handle, &g_timer_list);
 
-	if (callback->func) {
-		ret = callback->func(callback->cb_data);
-		/* It is true if eventloop_loop_stop is called in callback function. */
-		if (LOOP_IS_STOPPED(timer->loop)) {
-			return;
+	ret = eventloop_job_empty(handle, &is_empty);
+	if (ret != OK) {
+		return;
+	}
+
+	/* Add handle to a queue of jobs */
+	ret = eventloop_push_job(handle);
+	if (ret == OK && is_empty) {
+		/* Send signal to notify event */
+		ret = kill(handle->pid, SIGEL_EVENT);
+		elvdbg("[%d] watchdog kill %d\n", getpid(), handle->pid);
+		if (ret == OK) {
+		} else {
+			eldbg("Fail to send signal, pid %d errno %d\n", handle->pid, errno);
 		}
-		/* If callback function returns EVENTLOOP_CALLBACK_STOP, close and unregister the timer. */
-		if (!timer->repeat || ret == EVENTLOOP_CALLBACK_STOP) {
-			uv_close((uv_handle_t *)timer, (uv_close_cb)eventloop_unregister_timer);
+	}
+
+	/* Wake up task */
+	if (eventloop_get_state(handle->pid) == EL_STATE_WAIT_WAKEUP) {
+		elvdbg("[%d] SEND WAKEUP [%d]\n", getpid(), handle->pid);
+		ret = kill(handle->pid, SIGEL_WAKEUP);
+		if (ret != OK) {
+			eldbg("Fail to send signal, errno %d\n", errno);
 			return;
 		}
 	}
 }
 
-static el_timer_t *add_timer(el_loop_t *loop, unsigned int timeout, bool repeat, timeout_callback func, void *data)
+void eventloop_clean_timer_handle(el_handle_t *handle)
 {
 	int ret;
 	el_timer_t *timer;
-	timer_cb_t *callback;
-
-	if (loop == NULL || func == NULL) {
-		return NULL;
-	}
 
-	timer = (el_timer_t *)EL_ALLOC(sizeof(el_timer_t));
-	if (timer == NULL) {
-		eldbg("Failed to allocate timer\n");
-		return NULL;
+	if (handle == NULL) {
+		return;
 	}
 
-	callback = (timer_cb_t *)EL_ALLOC(sizeof(timer_cb_t));
-	if (callback == NULL) {
-		eldbg("Failed to allocate callback\n");
-		EL_FREE(timer);
-		return NULL;
-	}
-	callback->func = func;
-	callback->cb_data = data;
-	timer->data = (void *)callback;
+	timer = (el_timer_t *)&handle->data.timer;
 
-	ret = eventloop_register_timer(timer);
+	/* Clean timer handle */
+	ret = wd_delete(timer->wdog);
 	if (ret != OK) {
-		EL_FREE(callback);
-		EL_FREE(timer);
-		return NULL;
+		eldbg("Failed to delete watchdog\n");
+		return;
 	}
+	timer->wdog = NULL;
 
-	uv_update_time(loop);
-	uv_timer_init(loop, timer);
+	eventloop_decrease_app_handle_cnt(handle->pid);
 
-	if (!repeat) {
-		/* Add timer to be called once after timeout */
-		uv_timer_start(timer, (uv_timer_cb)timeout_callback_func, timeout, 0);
-	} else {
-		/* Add timer to be called every timeout repeatly */
-		uv_timer_start(timer, (uv_timer_cb)timeout_callback_func, timeout, timeout);
-	}
+	EL_FREE(handle);
+}
 
-	return timer;
+/* Remove all timers that task/pthread 'pid' registered */
+void eventloop_clean_app_timers(void)
+{
+	el_handle_t *rem_handle;
+	el_handle_t *handle;
+
+	handle = (el_handle_t *)sq_peek(&g_timer_list);
+	while (handle != NULL) {
+		if (handle->pid == getpid() && handle->mode == EL_MODE_DEFAULT) {
```

Comment:
how about changing the checking sequence between pid and mode? It will be faster when false case checking